client/starwhale/core/dataset/type.py

```diff
@@ -529,17 +541,38 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"Link uri:{self.uri}, offset:{self.offset}, size:{self.size}, data type:{self.data_type}, with localFS data:{self.with_local_fs_data}"
 
+    @http_retry
     def to_bytes(self, dataset_uri: t.Union[str, URI]) -> bytes:
+        from .store import ObjectStore
+
+        if self.signed_uri:
+            r = requests.get(self.signed_uri, timeout=10)
+            return r.content
         # TODO: auto inject dataset_uri in the loader process
         if isinstance(dataset_uri, str):
             dataset_uri = URI(dataset_uri, expected_type=URIType.DATASET)
 
         auth_name = self.auth.name if self.auth else ""
         if dataset_uri.instance_type == InstanceType.CLOUD:
-            store = ObjectStore.to_signed_http_backend(dataset_uri, auth_name)
+            key_compose = self, 0, 0
+            store = ObjectStore.to_signed_http_backend(dataset_uri)
         else:
-            store = ObjectStore.from_data_link_uri(self.uri, auth_name)
-        key_compose = self.local_fs_uri or self.uri, 0, 0
+            _up = urlparse(self.uri)
+            if _up.scheme:
+                key_compose = (
+                    Link(self.local_fs_uri) if self.local_fs_uri else self,
+                    0,
+                    0,
+                )
+                store = ObjectStore.from_data_link_uri(key_compose[0], auth_name)
+            else:
+                key_compose = (
+                    Link(self.local_fs_uri) if self.local_fs_uri else self,
+                    0,
+                    -2,
```

Comment:
could we use `0` as the end value of the key_compose?

Reply:
when the default value for end is `0` , `LocalFSStorageBackend._make_file`  will only read `1` byte in default case.
```python
class LocalFSStorageBackend(StorageBackend):
    def __init__(self) -> None:
        super().__init__(kind=SWDSBackendType.LocalFS)

    def _make_file(
        self, bucket: str, key_compose: t.Tuple[Link, int, int]
    ) -> FileLikeObj:
        _key_l, _start, _end = key_compose
        _key = _key_l.uri
        bucket_path = (
            Path(bucket).expanduser() if bucket.startswith("~/") else Path(bucket)
        )
        # TODO: tune reopen file performance, merge files
        data_path = bucket_path / _key[: DatasetStorage.short_sign_cnt]
        if not data_path.exists():
            data_path = bucket_path / _key

        with data_path.open("rb") as f:
            f.seek(_start)
            return io.BytesIO(f.read(_end - _start + 1))
```